### ISSUE TITLE
Cannot open include file: 'pthread.h'

### ISSUE THREAD

**Issue by daprose-dotcom** [mixxxdj/mixxx#11596]
### Bug Description

While compiling, I get this error message. How do I solve this problem?

fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory

### Version

2.3.5

### OS

Windows 10, Visual Studio Community 2019

**Comment by ronso0** [comment]
Not a Mixxx issue.

**Comment by daschuer** [comment]
@ikr83 has reported the same issue here: https://github.com/mixxxdj/mixxx/issues/11621#issuecomment-1576487778 Even if we know that it works if all is "correctly" configured it seems to be hard to find out what is correct. 
Maybe it is possible to fail with a better message or make is work instantly.

**Comment by daschuer** [comment]
Related: 
https://github.com/mixxxdj/mixxx/issues/11597

**Comment by ikr83** [comment]
I followed instructions here ...

https://github.com/mixxxdj/mixxx/wiki/Compiling%20on%20Windows
https://github.com/mixxxdj/mixxx/issues/11597

... and it works even if I didn't understand everything.

**Comment by daschuer** [comment]
Fine, now it is working. Can you reproduce your original issue? My idea is to make it initially work for new contributors or at least give less obscure error messages.

**Comment by ikr83** [comment]
The biggest difference is that I used CMake directly from VS2019 instead of using the command line method.

To be more specific:

At the beginning,

Download Visual Studio Community
https://aka.ms/vs/16/release/vs_community.exe

![1-Visual Studio Install Detail](https://github.com/mixxxdj/mixxx/assets/135498842/5d3bd376-6d23-4105-9cc3-64c4bb7ebd8c)
I didn't start VS2019 after installation.

Download Rust + Install
https://static.rust-lang.org/rustup/dist/x86_64-pc-windows-msvc/rustup-init.exe

Installation sccache
CMD + cargo install --git https://github.com/Be-ing/sccache.git --branch fix_msvc_fp

Download Mixxx source code
GitHub mixxxdj/mixxx

I tried with this code first:
https://github.com/mixxxdj/mixxx/releases/tag/2.3.5

Then I tried with this code:
https://github.com/mixxxdj/mixxx/archive/refs/heads/main.zip

Same results at the end...

Download + UNZIP code

Download buildenv
x64 Native Tools Command Prompt for VS 2019

    mixxx/tools/windows_buildenv.bat

Visual Studio

Open a local folder "mixxx folder"

Change the configuration and wait for CMake to finish each time:
x_64 legacy
x_64 native
x_64 off
x_64 portable

Press CTRL+SHIFT+B to generate everything.

After that, I can load Mixxx/build from the portable folder, but I can't find the Windows installer. The portable version of Mixxx works with my default preinstalled settings folder.

Originally, everything started from bug #11621.
So, I tried a third time after I overwrote the autodjprocessor.cpp in the Mixxx 2.3.5 code source folder.
https://github.com/mixxxdj/mixxx/blob/c03d885e975d342f38dae22efde7b2cd21d5a3d9/src/library/autodj/autodjprocessor.cpp
I regenerated to test and it works without the bug #11621. :)

During CMake, I saw error messages appearing, but they did not halt the generation process. However, only the "portable" folder contains a mixxx.exe executable. Maybe this is normal, I'm not sure...

**Comment by ikr83** [comment]
In build/portable some folders are missing. There is juste the script folder in fact

Few dll doesn't appear into my build/portable folder but appear in my "program files"

![DLL](https://github.com/mixxxdj/mixxx/assets/135498842/cd0417bb-e3a8-4a28-97d9-b12ecd2a1a8e)

These dll appear into my build/portable folder but doesn't appear in my "program files"
Qt5Test.dll

**Comment by ikr83** [comment]
[x_64_portable cmake sortie.txt](https://github.com/mixxxdj/mixxx/files/11657663/x_64_portable.cmake.sortie.txt)

**Comment by daschuer** [comment]
The x_64_portable.cmake.sortie.txt looks good. 
It should fail when something is missing. It looks like this was not the case in you very fist attempt. 

Can you remember what was the issue? Maybe you could remember what you did and reproduce this in a fresh build directory? 

In your current build we have two minor issues:
* The installer can be generated manually:`cpack -G WIX`
Where did you expect to find these information? How did you original expect it should work? 
* If you plan to start contributing, it is mandatory to use git. So instead if unzipping an archive, follow 
https://github.com/mixxxdj/mixxx/wiki/Bugfix-Workflow

**Comment by ikr83** [comment]
**What didn't work for me

Use Windows Command line (instead of VS2019) + don't install rust + install sccache with scoop

**What worked better

Download Rust + Install
https://static.rust-lang.org/rustup/dist/x86_64-pc-windows-msvc/rustup-init.exe

Installation sccache
CMD + cargo install --git https://github.com/Be-ing/sccache.git --branch fix_msvc_fp

Use VS2019 to build

********************

First time I build, so I didn't expect anything ... I knew cpack is necessary to generate a package but I didn't know how to use it.
Now I know the theory. :)

I'm not ready to start contributing for the moment but I will because I want to add the option to use a custom settings folder.

If during the installation of Mixxx, I can choose the folder where I place my configuration files, then I will place them in a folder connected to the cloud so that I can use Mixxx with different computers. Currently, I use robocopy and it works well, but it's still a makeshift solution. 

Next Episode... :)

Thanks a lot, I'll read the Bugfix Workflow.

**Comment by daschuer** [comment]
Ok, thank you. I will have a look how to improve this. 
A early error when you are not in the right terminal will help a lot I think. 

>  I want to add the option to use a custom settings folder.

This can already be done with a modified link on the desktop. You can add there any '--settingsPath` not sure if the database file itself works from a cloud though. 

A solution that makes this feature more discoverable would be a nice addition.

**Comment by ikr83** [comment]
Hi,

I haven't given up, I retried everything and I still can't build.

I reinstalled a clean environment
Visual Studio 2019 (C++) + Python + Git + Rustup + Sccache
I'm sharing a video of my command line procedure.
I don't see why it's not working...

https://github.com/mixxxdj/mixxx/assets/135498842/4fa14b5e-9b9d-4684-972c-86c25c1b590c

:/

**Comment by ikr83** [comment]
[CMakeError.log](https://github.com/mixxxdj/mixxx/files/11680283/CMakeError.log)

**Comment by ikr83** [comment]
If a make a second "cmake .." just after the first one, there is less errors.
[CMakeError.log](https://github.com/mixxxdj/mixxx/files/11680306/CMakeError.log)

**Comment by ikr83** [comment]
Here's a CHATGPT analysis of the specific errors:

    \src\controllers\bulk\bulkcontroller.cpp(55): error C2061: syntax error: identifier 'u_int8_t'
    This error indicates a syntax problem with the identifier 'u_int8_t'. Make sure you have included the necessary headers to declare this type.

    \src\controllers\bulk\bulkcontroller.cpp(58): error C2065: 'id': undeclared identifier
    This means that the variable 'id' has not been declared before it is being used. Check if you have properly declared and initialized all the required variables.

    \src\controllers\bulk\bulkcontroller.cpp(69): error C2660: 'get_string': function does not take 2 arguments
    This error suggests that you are calling the 'get_string' function with an incorrect number of arguments. Make sure you are calling the function with the appropriate arguments according to its signature.

    \src\controllers\bulk\bulkcontroller.cpp(68): error C2512: 'Controller::Controller': no appropriate default constructor available
    This error indicates that there is a missing appropriate default constructor for the 'Controller' class. Ensure that you have defined a default constructor for this class or use an existing constructor correctly.

    \src\controllers\bulk\bulkcontroller.cpp(78): error C2660: 'get_string': function does not take 2 arguments
    This error is similar to error 3. Make sure you are calling the 'get_string' function correctly with the required arguments.

**Comment by daschuer** [comment]
Please pull my patch and check if anything else is failing: 
``` 
git pull https://github.com/daschuer/mixxx.git u_int
``

**Comment by ikr83** [comment]
I fix u_int8_t, the build completes successfully, I start mixxx.exe and it works!

:D

***

Next step,
I run cpack. cpack alerts me that NSIS is missing, so I install NSIS. I restart cpack.

In VS2019
CPack Error: Problem running NSIS command: "C:/Program Files (x86)/NSIS/makensis.exe" "C:/Users/david/Desktop/mixxx/build/_CPack_Packages/win64/NSIS/project.nsi"
Please check C:/Users/david/Desktop/mixxx/build/_CPack_Packages/win64/NSIS/NSISOutput.log for errors
CPack Error: Problem compressing the directory
CPack Error: Error when generating package: mixxx

In NSISOutput.log
File: "C:/Users/david/Desktop/mixxx/res/images/mixxx_install_logo.bmp" -> no files found.
Usage: File [/nonfatal] [/a] ([/r] [/x filespec [...]] filespec [...] |
   /oname=outfile one_file_only)
Error in macro MUI_HEADERIMAGE_INIT on macroline 25
Error in macro MUI_GUIINIT on macroline 3
Error in macro MUI_FUNCTION_GUIINIT on macroline 4
Error in macro MUI_INSERT on macroline 11
Error in macro MUI_LANGUAGE on macroline 14
Error in script "C:/Users/david/Desktop/mixxx/build/_CPack_Packages/win64/NSIS/project.nsi" on line 571 -- aborting creation process

I can see mixxx_install_logo.bmp in my explorer...

project.nsi" on line 571
!insertmacro MUI_LANGUAGE "English" ;first language is the default language

***

I delete the _CPack_Packages folder, and I remember you had advised me to use "cpack -G WIX," but I don't have WIX installed.
I install .NET SDK (7.0).
I install Wix.
dotnet tool install --global wix.
I verify by checking the version: wix --version.
It's okay (4.0.1+6e6eb478).

CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/Internal/CPack/CPackWIX.cmake:12 (message):
  Could not find the WiX candle executable.

I run cpack -G WIX

CPack Error: Error while executing CPackWIX.cmake
CPack Error: Fatal WiX Generator Error
CPack Error: Problem compressing the directory
CPack Error: Error when generating package: mixxx

I think there is a problem with new Wix Version
https://wixtoolset.org/docs/fourthree/
"A lot about WiX has changed between v3 and v4"

"For command-line afficionados, most executables have been merged into a single wix.exe tool with commands. For example, in WiX v3, you might build by calling Candle.exe one or more times to compile your authoring and then calling Light.exe to link and bind the compiled authoring into an .msi package. In WiX v4 using wix.exe, that's one command: wix build -o product.msi product.wxs."

:/

**Comment by JoergAtGithub** [comment]
> The issue with "u_int8_t" looks like a typo. I have no idea why our CI did not detect it.

This is because this affects only bulkcontroller.cpp and the Windows CI builds are with BULK=OFF.

**Comment by daschuer** [comment]
Is that correct? I think we should as as align our home builds with the CI defaults.

**Comment by daschuer** [comment]
> cpack alerts me that NSIS

Another pitfall, sorry. 

Thank you for all the hints. I will take a look.

**Comment by JoergAtGithub** [comment]
Yes, because BULK USB can't be used for security reasons without installing additional drivers on Windows.
Th CMakeSettings.txt and CI are in sync. If you compile from command line, you need to specify these yourself.
The real solution is using CMake Presets instead of CMakeSettings.txt, because these can be used from commandline, CI and IDE builds and work accross all platforms.

**Comment by daschuer** [comment]
Presets are very welcome to build different configurations on the same host. This allows a lot more than just "Debug" and Release. We can also auto generate a CMakeUserPreset.json like we currently do with "CMakeSettings.json"

Apart CMake presets, we should also choose reasonable defaults when starting cmake without "--preset"
I think in this case here we need switch off BULK by default.

**Comment by ikr83** [comment]
Sorry for using the same page to address all my little issues to achieve the build+pack... but since I started, I feel obligated to continue...  Regarding my "cpack issue", changing "u_int8_t" to "uint8_t" was enough... Tonight, I will uninstall Wix 4, install Wix 3, and see if it works better.

**Comment by daschuer** [comment]
This is issue is the perfect place for such things. I expect that there are are number of users suffer like you, but instead of analyzing the issue they give up.

**Comment by ikr83** [comment]
Oki!

cpack -G WIX with Wix4 doesn't work for me ( because there is the candle issue https://wixtoolset.org/docs/fourthree/ ).
cpack -G WIX works better with WiX Toolset v3.11

![Finally](https://github.com/mixxxdj/mixxx/assets/135498842/5db1d5f3-fa1c-4686-9c29-11fc09a52923)

Thank you very much!

**Comment by ikr83** [comment]
To resume,

Windows environment setup

To build
Install Visual Studio 2019 Community

![Visual Studio 2019 Community](https://github.com/mixxxdj/mixxx/assets/135498842/d8bfc6ba-36bc-46f4-a6e3-1e021ceaa1b2)

Optionnal
Installation of Rustup with the installer
https://static.rust-lang.org/rustup/dist/x86_64-pc-windows-msvc/rustup-init.exe
Installation of Sccache with cargo (Rustup)
cargo install --git https://github.com/Be-ing/sccache.git --branch fix_msvc_fp

To make an installer after building
Install WiX Toolset v3.11 (Wix 4 doesn't work because there is a candle issue https://wixtoolset.org/docs/fourthree/)

Run x64 Native Tools Command Prompt for VS 2019

cd c:/users/david/Desktop
git clone https://github.com/mixxxdj/mixxx.git
cd mixxx/tools
windows_buildenv.bat

corrections
#11596 mixxx\src\controllers\bulk\bulkcontroller.cpp (this one is necessary for the building)
#11621 mixxx\src\library\autodj\autodjprocessor.cpp (this one is about the autodj transition isue)

cd ../build
cmake ..
cmake --build .
cpack -G WIX

:D

**Comment by daschuer** [comment]
This means we are done, once the pending PRs are merged? 

I am sure the scoop installation of sccache will also work.
Does this mean less hassle compared to the cargo installation? Does cargo build sccache from the source? 

Maybe we should describe the scoop way as recommended. @JoergAtGithub what do you think? 

sccache is by the way optional, it is strongly recommend for contributors that build different branches though.
For a single branch it has no advantages.

**Comment by ikr83** [comment]
Indeed, when the source code is up to date, it will be fine! But be careful with the cpack -G WIX command, which requires the installation of WIX v3. The latest version of WIX (v4) doesn't work (candle issue). I don't think Rustup installs from source code (the same goes for Scoop).
